python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,261 @@
+            raise WorkflowEventHandleError(workflow_id, f"The same {event_key} is used to get payload again.")
+
+        self.event_key_payload[workflow_id][event_key] = asyncio.Future()
+        return await self.event_key_payload[workflow_id][event_key]
+
+    async def report_checkpointed(self, workflow_id, event_key, confirmation) -> str:
+        """Internal method used by HTTPListner to confirm the received event has been checkpointed by workflow
+        """
+        try:
+            self.event_checkpoint_pending[event_key].set_result(confirmation)
+        except KeyError:
+            logger.error(f"{event_key} cannot be found to acknowledge request. "
+                         f"The event provider may have been restarted."
+            )
+            raise WorkflowEventHandleError(workflow_id, f"{event_key} cannot be found to acknowledge request.")
+        return "OK"
+
+def get_or_create_http_event_provider() -> "ActorHandle":
+    """get_or_create_http_event_provider returns a handle to an HTTPEventProvider Serve deployment,
+    name defined as common.HTTP_EVENT_PROVIDER_NAME, if available. If not, it creates such a deployment
+    and returns a handle to it.
+    """
+    try:
+        return ray.serve.get_deployment(common.HTTP_EVENT_PROVIDER_NAME).get_handle(sync=True)
+    except:
+        # HTTPEventProvider Serve endpoint was not found. Create an deployment.
+        HTTPEventProvider.deploy()
+        return ray.serve.get_deployment(common.HTTP_EVENT_PROVIDER_NAME).get_handle(sync=True)
+
+class HTTPListener(EventListener):
+    """HTTPLister is defined to work with the HTTPEventProvider. It implements two
+    APIs, poll_for_event() and event_checkpointed(). An instance of HTTPListener will
+    be started by the get_message() of the workflow.wait_for_event() to listen for an event
+    from the HTTPEventProvider instance (a Ray Serve deployment). Another instance of HTTPListener
+    will be started by the message_committed() of the workflow.wait_for_event() to confirm
+    that the event has been checkpointed.
+
+
+    Example definition
+    ==================
+
+    ```
+    class HTTPListener:
+
+        def __init__(self):
+            self.handle = get_or_create_http_event_provider()
+
+        async def poll_for_event(self, event_key) -> Event:
+
+        async def event_checkpointed(self, event) -> None:
+
+    ```
+
+    Example Usage
+    =============
+    >>> from ray.workflow.http_event_provider import HTTPEventProvider, HTTPListener
+    >>> ray.init(address='auto', namespace='serve')
+    >>> serve.start(detached=True)
+    >>> event_node = workflow.wait_for_event( # doctest: +SKIP
+    ...     HTTPListener, event_key='')
+    >>> handle_event = ... # doctest: +SKIP
+    >>> handle_event.bind(event_node).run() # doctest: +SKIP
+    >>>
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.handle = get_or_create_http_event_provider()
```

Comment:
I think we should put it into WAM to avoid race condition.

Reply:
Done